lib/ngx/ssl.lua

```diff
@@ -17,53 +17,149 @@ local get_string_buf = base.get_string_buf
+    ngx_lua_ffi_free_cert = C.ngx_http_lua_ffi_free_cert
+    ngx_lua_ffi_free_priv_key = C.ngx_http_lua_ffi_free_priv_key
+
+elseif subsystem == 'stream' then
+    ffi.cdef[[
+    int ngx_stream_lua_ffi_ssl_set_der_certificate(ngx_stream_lua_request_t *r,
+        const char *data, size_t len, char **err);
+
+    int ngx_stream_lua_ffi_ssl_clear_certs(ngx_stream_lua_request_t *r,
+        char **err);
+
+    int ngx_stream_lua_ffi_ssl_set_der_private_key(ngx_stream_lua_request_t *r,
+        const char *data, size_t len, char **err);
+
+    int ngx_stream_lua_ffi_ssl_raw_server_addr(ngx_stream_lua_request_t *r,
+        char **addr, size_t *addrlen, int *addrtype, char **err);
+
+    int ngx_stream_lua_ffi_ssl_server_name(ngx_stream_lua_request_t *r,
+        char **name, size_t *namelen, char **err);
+
+    int ngx_stream_lua_ffi_ssl_raw_client_addr(ngx_stream_lua_request_t *r,
+        char **addr, size_t *addrlen, int *addrtype, char **err);
+
+    int ngx_stream_lua_ffi_cert_pem_to_der(const unsigned char *pem,
+        size_t pem_len, unsigned char *der, char **err);
+
+    int ngx_stream_lua_ffi_priv_key_pem_to_der(const unsigned char *pem,
+        size_t pem_len, unsigned char *der, char **err);
+
+    int ngx_stream_lua_ffi_ssl_get_tls1_version(ngx_stream_lua_request_t *r,
+        char **err);
+
+    void *ngx_stream_lua_ffi_parse_pem_cert(const unsigned char *pem,
+        size_t pem_len, char **err);
+
+    void *ngx_stream_lua_ffi_parse_pem_priv_key(const unsigned char *pem,
+        size_t pem_len, char **err);
+
+    int ngx_stream_lua_ffi_set_cert(void *r, void *cdata, char **err);
+
+    int ngx_stream_lua_ffi_set_priv_key(void *r, void *cdata, char **err);
+
+    void ngx_stream_lua_ffi_free_cert(void *cdata);
+
+    void ngx_stream_lua_ffi_free_priv_key(void *cdata);
+    ]]
+
+    ngx_lua_ffi_ssl_set_der_certificate =
+        C.ngx_stream_lua_ffi_ssl_set_der_certificate
+    ngx_lua_ffi_ssl_clear_certs = C.ngx_stream_lua_ffi_ssl_clear_certs
+    ngx_lua_ffi_ssl_set_der_private_key =
+        C.ngx_stream_lua_ffi_ssl_set_der_private_key
+    ngx_lua_ffi_ssl_raw_server_addr = C.ngx_stream_lua_ffi_ssl_raw_server_addr
+    ngx_lua_ffi_ssl_server_name = C.ngx_stream_lua_ffi_ssl_server_name
+    ngx_lua_ffi_ssl_raw_client_addr = C.ngx_stream_lua_ffi_ssl_raw_client_addr
+    ngx_lua_ffi_cert_pem_to_der = C.ngx_stream_lua_ffi_cert_pem_to_der
+    ngx_lua_ffi_priv_key_pem_to_der = C.ngx_stream_lua_ffi_priv_key_pem_to_der
+    ngx_lua_ffi_ssl_get_tls1_version = C.ngx_stream_lua_ffi_ssl_get_tls1_version
+    ngx_lua_ffi_parse_pem_cert = C.ngx_stream_lua_ffi_parse_pem_cert
+    ngx_lua_ffi_parse_pem_priv_key = C.ngx_stream_lua_ffi_parse_pem_priv_key
+    ngx_lua_ffi_set_cert = C.ngx_stream_lua_ffi_set_cert
+    ngx_lua_ffi_set_priv_key = C.ngx_stream_lua_ffi_set_priv_key
+    ngx_lua_ffi_free_cert = C.ngx_stream_lua_ffi_free_cert
+    ngx_lua_ffi_free_priv_key = C.ngx_stream_lua_ffi_free_priv_key
+end
```

Comment:
style: there used to be 2 blank lines below this, and now a single one only

Reply:
Fixed.